lxd/api_cluster.go

```diff
@@ -753,30 +753,37 @@ func clusterPutJoin(d *Daemon, r *http.Request, req api.ClusterPut) response.Res
 			return err
 		}
 
-		// Connect to MAAS
-		url, key := s.GlobalConfig.MAASController()
-		machine := nodeConfig.MAASMachine()
-		err = d.setupMAASController(url, key, machine)
+		d.globalConfigMu.Lock()
+		d.localConfig = nodeConfig
+		d.globalConfigMu.Unlock()
+
+		// Get the current (updated) config.
+		var existingClusterConfig *clusterConfig.Config
+		err = d.db.Cluster.Transaction(context.TODO(), func(ctx context.Context, tx *db.ClusterTx) error {
+			existingClusterConfig, err = clusterConfig.Load(ctx, tx)
+			if err != nil {
+				return err
+			}
+
+			return nil
+		})
 		if err != nil {
 			return err
 		}
 
-		// Handle external authentication/RBAC
-		candidAPIURL, candidAPIKey, candidExpiry, candidDomains := s.GlobalConfig.CandidServer()
-		rbacAPIURL, rbacAPIKey, rbacExpiry, rbacAgentURL, rbacAgentUsername, rbacAgentPrivateKey, rbacAgentPublicKey := s.GlobalConfig.RBACServer()
+		d.globalConfigMu.Lock()
+		d.globalConfig = existingClusterConfig
```

Comment:
Shall we update `d.localConfig = nodeConfig` here inside this lock too and save a lock as well as avoiding doing a partial update to only `d.localConfig` above in case of error?

Reply:
Yeah that makes sense.

Reply:
Done.